File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/response/model/node/NodeKeyDetails.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.retrofit.response.model.node;

import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;
import android.util.Patterns;

import com.flowcrypt.email.model.PgpContact;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.mail.internet.AddressException;
import javax.mail.internet.InternetAddress;

/**
 * @author Denis Bondarenko
 * Date: 2/11/19
 * Time: 1:23 PM
 * E-mail: DenBond7@gmail.com
 */
public class NodeKeyDetails implements Parcelable {

  public static final Creator<NodeKeyDetails> CREATOR = new Creator<NodeKeyDetails>() {
    @Override
    public NodeKeyDetails createFromParcel(Parcel source) {
      return new NodeKeyDetails(source);
    }

    @Override
    public NodeKeyDetails[] newArray(int size) {
      return new NodeKeyDetails[size];
    }
  };

  @Expose
  private Boolean isDecrypted;

  @Expose
  @SerializedName("private")
  private String privateKey;

  @Expose
  @SerializedName("public")
  private String publicKey;

  @Expose
  private List<String> users;

  @Expose
  private List<KeyId> ids;

  @Expose
  private long created;

  @Expose
  private Algo algo;

  public NodeKeyDetails() {
  }

  protected NodeKeyDetails(Parcel in) {
    this.isDecrypted = (Boolean) in.readValue(Boolean.class.getClassLoader());
    this.privateKey = in.readString();
    this.publicKey = in.readString();
    this.users = in.createStringArrayList();
    this.ids = in.createTypedArrayList(KeyId.CREATOR);
    this.created = in.readLong();
    this.algo = in.readParcelable(Algo.class.getClassLoader());
  }

  @Override
  public int describeContents() {
    return 0;
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    dest.writeValue(this.isDecrypted);
    dest.writeString(this.privateKey);
    dest.writeString(this.publicKey);
    dest.writeStringList(this.users);
    dest.writeTypedList(this.ids);
    dest.writeLong(this.created);
    dest.writeParcelable(this.algo, flags);
  }

  public String getPrivateKey() {
    return privateKey;
  }

  public String getPublicKey() {
    return publicKey;
  }

  public List<String> getUsers() {
    return users;
  }

  public List<KeyId> getIds() {
    return ids;
  }

  public long getCreated() {
    return TimeUnit.MILLISECONDS.convert(created, TimeUnit.SECONDS);
  }

  public Algo getAlgo() {
    return algo;
  }

  public PgpContact getPrimaryPgpContact() {
    KeyId keyId = ids.get(0);
    String email = null;
    String name = null;
    try {
      InternetAddress[] internetAddresses = InternetAddress.parse(users.get(0));
      email = internetAddresses[0].getAddress();
      name = internetAddresses[0].getPersonal();
    } catch (AddressException e) {
      e.printStackTrace();

      //It means the input string has a wrong format. So we need to try to extract only an email address using regex.
      Pattern pattern = Patterns.EMAIL_ADDRESS;
      Matcher matcher = pattern.matcher(users.get(0));
      if (matcher.find()) {
        email = matcher.group();
        name = email;
      }
    }

    return new PgpContact(email, name, publicKey, !TextUtils.isEmpty(publicKey), null, false,
        keyId.getFingerprint(), keyId.getLongId(), keyId.getKeywords(), 0);
  }

  public ArrayList<PgpContact> getPgpContacts() {
    ArrayList<PgpContact> pgpContacts = new ArrayList<>();

    for (String user : users) {
      try {
        InternetAddress[] internetAddresses = InternetAddress.parse(user);

        for (InternetAddress internetAddress : internetAddresses) {
          String email = internetAddress.getAddress();
          String name = internetAddress.getPersonal();

          pgpContacts.add(new PgpContact(email, name));
        }
      } catch (AddressException e) {
        e.printStackTrace();
      }
    }

    return pgpContacts;
  }

  public String getLongId() {
    return ids.get(0).getLongId();
  }

  public String getFingerprint() {
    return ids.get(0).getFingerprint();
  }

  public String getKeywords() {
    return ids.get(0).getKeywords();
  }

  public boolean isPrivate() {
    return !TextUtils.isEmpty(privateKey);
  }

  public Boolean isDecrypted() {
    return isDecrypted;
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/response/model/node/NodeKeyDetails.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors:
 *  DenBond7
 *  Ivan Pizhenko
 */

package com.flowcrypt.email.api.retrofit.response.model.node

import android.os.Parcel
import android.os.Parcelable
import android.text.TextUtils
import android.util.Patterns
import com.flowcrypt.email.database.entity.AccountEntity
import com.flowcrypt.email.database.entity.KeyEntity
import com.flowcrypt.email.model.PgpContact
import com.flowcrypt.email.security.model.PrivateKeySourceType
import com.flowcrypt.email.util.exception.FlowCryptException
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import java.util.*
import javax.mail.internet.AddressException
import javax.mail.internet.InternetAddress

/**
 * @author Denis Bondarenko
 * Date: 2/11/19
 * Time: 1:23 PM
 * E-mail: DenBond7@gmail.com
 */
data class NodeKeyDetails constructor(@Expose val isFullyDecrypted: Boolean?,
                                      @Expose val isFullyEncrypted: Boolean?,
                                      @Expose @SerializedName("private") val privateKey: String?,
                                      @Expose @SerializedName("public") val publicKey: String?,
                                      @Expose val users: List<String>?,
                                      @Expose val ids: List<KeyId>?,
                                      @Expose val created: Long,
                                      @Expose val lastModified: Long,
                                      @Expose val expiration: Long,
                                      @Expose val algo: Algo?,
                                      var passphrase: String?,
                                      var errorMsg: String?) : Parcelable {

  val primaryPgpContact: PgpContact
    get() = determinePrimaryPgpContact()
  val pgpContacts: ArrayList<PgpContact>
    get() = determinePgpContacts()
  val fingerprint: String?
    get() = ids?.first()?.fingerprint
  val isPrivate: Boolean
    get() = !TextUtils.isEmpty(privateKey)

  val isExpired: Boolean
    get() = expiration > 0 && (System.currentTimeMillis() / 1000 > expiration)

  val mimeAddresses: List<InternetAddress>
    get() = parseMimeAddresses()

  val isPartiallyEncrypted: Boolean
    get() {
      return isFullyDecrypted == false && isFullyEncrypted == false
    }

  constructor(source: Parcel) : this(
      source.readValue(Boolean::class.java.classLoader) as Boolean?,
      source.readValue(Boolean::class.java.classLoader) as Boolean?,
      source.readString(),
      source.readString(),
      source.createStringArrayList(),
      source.createTypedArrayList(KeyId.CREATOR),
      source.readLong(),
      source.readLong(),
      source.readLong(),
      source.readParcelable<Algo>(Algo::class.java.classLoader),
      source.readString(),
      source.readString()
  )

  override fun describeContents() = 0

  override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
    writeValue(isFullyDecrypted)
    writeValue(isFullyEncrypted)
    writeString(privateKey)
    writeString(publicKey)
    writeStringList(users)
    writeTypedList(ids)
    writeLong(created)
    writeLong(lastModified)
    writeLong(expiration)
    writeParcelable(algo, 0)
    writeString(passphrase)
    writeString(errorMsg)
  }

  private fun determinePrimaryPgpContact(): PgpContact {
    val address = users?.first()

    address?.let {
      val fingerprintFromKeyId = ids?.first()?.fingerprint
      var email: String? = null
      var name: String? = null
      try {
        val internetAddresses = InternetAddress.parse(it)
        email = internetAddresses.first().address
        name = internetAddresses.first().personal
      } catch (e: AddressException) {
        e.printStackTrace()
        val pattern = Patterns.EMAIL_ADDRESS
        val matcher = pattern.matcher(users!!.first())
        if (matcher.find()) {
          email = matcher.group()
          name = email
        }
      }

      if (email == null) {
        throw object : FlowCryptException("No user ids with mail address") {}
      }

      return PgpContact(
          email = email.toLowerCase(Locale.US),
          name = name,
          pubkey = publicKey,
          hasPgp = !TextUtils.isEmpty(publicKey),
          client = null,
          fingerprint = fingerprintFromKeyId
      )
    }

    return PgpContact("", "")
  }

  private fun determinePgpContacts(): ArrayList<PgpContact> {
    val pgpContacts = ArrayList<PgpContact>()

    users?.let {
      for (user in it) {
        try {
          val internetAddresses = InternetAddress.parse(user)

          for (internetAddress in internetAddresses) {
            val email = internetAddress.address.toLowerCase(Locale.US)
            val name = internetAddress.personal

            pgpContacts.add(PgpContact(email, name))
          }
        } catch (e: AddressException) {
          e.printStackTrace()
        }
      }
    }

    return pgpContacts
  }

  private fun parseMimeAddresses(): List<InternetAddress> {
    val results = mutableListOf<InternetAddress>()

    for (user in users ?: emptyList()) {
      try {
        results.addAll(listOf(*InternetAddress.parse(user)))
      } catch (e: AddressException) {
        //do nothing
      }
    }

    return results
  }

  fun toKeyEntity(accountEntity: AccountEntity): KeyEntity {
    return KeyEntity(
        fingerprint = fingerprint
            ?: throw NullPointerException("nodeKeyDetails.fingerprint == null"),
        account = accountEntity.email.toLowerCase(Locale.US),
        accountType = accountEntity.accountType,
        source = PrivateKeySourceType.BACKUP.toString(),
        publicKey = publicKey?.toByteArray()
            ?: throw NullPointerException("nodeKeyDetails.publicKey == null"),
        privateKey = privateKey?.toByteArray()
            ?: throw NullPointerException("nodeKeyDetails.privateKey == null"),
        storedPassphrase = passphrase)
  }

  companion object {
    @JvmField
    val CREATOR: Parcelable.Creator<NodeKeyDetails> = object : Parcelable.Creator<NodeKeyDetails> {
      override fun createFromParcel(source: Parcel): NodeKeyDetails = NodeKeyDetails(source)
      override fun newArray(size: Int): Array<NodeKeyDetails?> = arrayOfNulls(size)
    }
  }
}
